### PR TITLE
RavenDB-16114 Group fanout performance hint into single dialog

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

File: src/Raven.Server/NotificationCenter/Indexing.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        private static readonly string Source = "Indexing";
+        private static readonly string IndexingTableId = PerformanceHint.GetKey(PerformanceHintType.Indexing, Source);
+
+        private readonly NotificationCenter _notificationCenter;
+        private readonly NotificationsStorage _notificationsStorage;
+        private readonly string _database;
+
+        private DateTime _warningUpdateTime = new DateTime();
+        private ConcurrentQueue<(string Action, WarnIndexOutputsPerDocument.WarningDetails warningDetails)> _warningQueue =
```
Comment (ml054) [inline src/Raven.Server/NotificationCenter/Indexing.cs:23]: didn't we talk to use dict for that and limit items per every index?

File: src/Raven.Server/NotificationCenter/Indexing.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        private void UpdateIndexing(object state)
+        {
+            try
+            {
+                if (_warningQueue.IsEmpty)
+                    return;
+
+                PerformanceHint indexingWarnings = null;
+
+                while (_warningQueue.TryDequeue(
```
Comment (ml054) [inline src/Raven.Server/NotificationCenter/Indexing.cs:75]: it removes other items until it finds matching one?

File: src/Raven.Server/NotificationCenter/Indexing.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            var now = SystemTime.UtcNow;
+            var update = _warningUpdateTime;
+
+            if (now - update < TimeSpan.FromSeconds(15))
+                return;
+
+            _warningUpdateTime = now;
+            
+            _indexOutputsWarning.LastWarningTime = now;
+            _warningQueue.Enqueue((indexName, _indexOutputsWarning));
```
Comment (ml054) [inline src/Raven.Server/NotificationCenter/Indexing.cs:49]: you put that in queue: new ConcurrentQueue<(string Action, WarnIndexOutputsPerDocument.WarningDetails warningDetails)>();

should rename paramter name Action -> IndexName?

File: src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts [hunk 4]
```diff
@@ -3,11 +3,101 @@ import abstractNotification = require("common/notifications/models/abstractNotif
...
+                    sortable: x => x.LastWarningTime
+                })
+            ];
+        });
+
+        this.columnPreview.install(".indexingDetails", ".js-indexing-details-tooltip",
+            (details: indexingDetailsItemDto, column: textColumn<indexingDetailsItemDto>, e: JQueryEventObject,
+             onValue: (context: any, valueToCopy?: string) => void) => {
+                const value = column.getCellValue(details);
+                if (column.header === "Last warned atDate") {
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts]: I don't see column with such name.

File: src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts [hunk 5]
```diff
@@ -3,11 +3,101 @@ import abstractNotification = require("common/notifications/models/abstractNotif
...
+                totalResultCount: this.tableItems.length
+            });
+    }
+
+    private mapItems(details: Raven.Server.NotificationCenter.Notifications.Details.WarnIndexOutputsPerDocument): indexingDetailsItemDto[] {
+        return _.flatMap(details.Warnings, (value, key) => {
+            return value.map(item =>
+                ({
+                    IndexName: key,
+                    NumberOfExceedingDocuments: item.NumberOfExceedingDocuments,
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts]: { IndexName: key, ...item }

Comment (ml054) [reply src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts]: + avoid using as operator

File: src/Raven.Studio/wwwroot/App/views/common/notificationCenter/detailViewer/performanceHint/indexingDetails.html [hunk 6]
```diff
@@ -14,37 +14,20 @@ <h3 class="modal-title" id="myModalLabel" data-bind="text: hint.title, attr:{ cl
...
-                    </div>
-                </div>
-                <div class="bg-info padding padding-sm ">
-                    <strong>Suggestion:</strong>
-                    <div class="small" data-bind="text: Suggestion"></div>
+            <div style="position: relative; height: 300px">
+                <virtual-grid style="height: 300px" class="resizable flex-window" params="controller: gridController"></virtual-grid>
+            </div>
+            <div class="bg-info text-info padding">
+                <strong data-bind="visible: suggestions.length = 1">Suggestion:</strong>
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/App/views/common/notificationCenter/detailViewer/performanceHint/indexingDetails.html]: you trim array length here! ;-)

File: src/Raven.Server/NotificationCenter/Indexing.cs (hunk 1 above)
Comment (ml054) [inline src/Raven.Server/NotificationCenter/Indexing.cs]: rename action -> index name

File: src/Raven.Server/NotificationCenter/Indexing.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+            if (_indexingTimer != null)
+                return;
+
+            lock (_locker)
+            {
+                if (_indexingTimer != null)
+                    return;
+
+                _indexingTimer = new Timer(UpdateIndexing, null, TimeSpan.FromSeconds(0), TimeSpan.FromMinutes(5));
```
Comment (ml054) [inline src/Raven.Server/NotificationCenter/Indexing.cs]: do we want zero for first execution timeout?

Comment (Danielle9897) [reply src/Raven.Server/NotificationCenter/Indexing.cs]: Leaving this - as discussed

File: src/Raven.Server/NotificationCenter/Indexing.cs (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Server/NotificationCenter/Indexing.cs]: done

File: src/Raven.Studio/wwwroot/App/views/common/notificationCenter/detailViewer/performanceHint/indexingDetails.html (hunk 6 above)
Comment (Danielle9897) [reply src/Raven.Studio/wwwroot/App/views/common/notificationCenter/detailViewer/performanceHint/indexingDetails.html]: done

File: src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts]: done

File: src/Raven.Server/NotificationCenter/Indexing.cs (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Server/NotificationCenter/Indexing.cs:49]: done

File: src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts (hunk 5 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/common/notificationCenter/detailViewer/performanceHint/indexingDetails.ts]: done

**Review by ml054 — APPROVED** [review]

**Pull request merged**